### PR TITLE
feature: support server to expose Nacos services by mounting SLB

### LINKED ISSUE #4687: support server to expose Nacos services by mounting SLB
## Why you need it?
 Is your feature request related to a problem? Please describe in details

In cloud architectures, middleware deployment and business deployment networks are often isolated for security and functional responsibilities. Currently, server supports the injection of -H or SEATA_IP, but it cannot meet the automatic deployment of containerized(K8S) deployment.

## How it could be?
A clear and concise description of what you want to happen. You can explain more about input of the feature, and output of it.

### PR DESCRIPTION
**Pull request by 2129zxl**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did
fixes #4687

### CODE REVIEW

**Comment by 2129zxl** [comment]
> 为什么不使用SEATA_IP或者-h？
该场景为server运行时动态绑定公网地址通过ingress暴露，非启动时注入，客户端通过配置项做公网地址的服务发现

**Comment by a364176773** [comment]
> > 为什么不使用SEATA_IP或者-h？
> > 该场景为server运行时动态绑定公网地址通过ingress暴露，非启动时注入，客户端通过配置项做公网地址的服务发现

server都在运行中了，服务已经暴露至nacos中，何以自动绑定成公网ip又通过k8s ingress暴露，且注册在nacos中的？可以详细讲下这个服务注册的流程吗？

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 1]
```diff
@@ -129,6 +141,25 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
         if (clusterName == null) {
             return null;
         }
+        Pattern nacosRegistryPatternForMeta = StringUtils.isBlank(getNamingProperties().getProperty(SERVICE_META_PATTERN))?
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:144]: create `temp properties`  for every  interact with TCs ?

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 2]
```diff
@@ -61,13 +67,19 @@ public class NacosRegistryServiceImpl implements RegistryService<EventListener>
     private static final String PASSWORD = "password";
     private static final String ACCESS_KEY = "accessKey";
     private static final String SECRET_KEY = "secretKey";
+    private static final String SERVICE_META_PATTERN = "pattern";
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:70]: need clear meaning

File: seata-spring-autoconfigure/seata-spring-autoconfigure-core/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java [hunk 3]
```diff
@@ -114,4 +115,15 @@ public RegistryNacosProperties setSecretKey(String secretKey) {
         this.secretKey = secretKey;
         return this;
     }
+
+    public String getPattern() {
```
Comment (slievrly) [inline seata-spring-autoconfigure/seata-spring-autoconfigure-core/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java:119]: how to use this configuration item，should also add in:
https://github.com/seata/seata/tree/develop/script/client/conf
https://github.com/seata/seata/tree/develop/script/client/spring
https://github.com/seata/seata/blob/develop/script/config-center/config.txt

**Review by slievrly — COMMENTED** [review]
Configuration item completion:
https://github.com/seata/seata/blob/develop/seata-spring-autoconfigure/seata-spring-autoconfigure-client/src/main/resources/META-INF/additional-spring-configuration-metadata.json

File: changes/zh-cn/develop.md [hunk 4]
```diff
@@ -3,6 +3,7 @@
 <!-- 请根据PR的类型添加 `变更记录` 到以下对应位置(feature/bugfix/optimize/test) 下 -->
 
 ### feature：
+  - [[#4676](https://github.com/seata/seata/pull/4676)] 新增nacos注册中心公网地址服务发现机制
```
Comment (slievrly) [inline changes/zh-cn/develop.md:6]: The commit log must have an exact meaning. What is the public network for open source users?

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 5]
```diff
@@ -129,6 +141,25 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
         if (clusterName == null) {
             return null;
         }
+        Pattern nacosRegistryPatternForMeta = StringUtils.isBlank(getNamingProperties().getProperty(SERVICE_META_PATTERN))?
+                DEFAULT_CUSTOMER_REGISTRY_PATTERN : Pattern.compile(getNamingProperties().getProperty(SERVICE_META_PATTERN));
+        if(nacosRegistryPatternForMeta.matcher(getNamingProperties().getProperty(PRO_SERVER_ADDR_KEY)).matches()){
+            LOGGER.info("visit seata on aliyun with pubnet mode");
+            if (!CLUSTER_ADDRESS_MAP.containsKey(PUBLIC_NAMING_ADDRESS_PREFIX+clusterName)){
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:148]: This code does not have format, can not pass the check style.

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 6]
```diff
@@ -129,6 +141,25 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
         if (clusterName == null) {
             return null;
         }
+        Pattern nacosRegistryPatternForMeta = StringUtils.isBlank(getNamingProperties().getProperty(SERVICE_META_PATTERN))?
+                DEFAULT_CUSTOMER_REGISTRY_PATTERN : Pattern.compile(getNamingProperties().getProperty(SERVICE_META_PATTERN));
+        if(nacosRegistryPatternForMeta.matcher(getNamingProperties().getProperty(PRO_SERVER_ADDR_KEY)).matches()){
+            LOGGER.info("visit seata on aliyun with pubnet mode");
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:147]: The concept of `pubnet` needs to be rethought. The word `aliyun` should not appear.

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 7]
```diff
@@ -129,6 +141,25 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
...
+            if (!CLUSTER_ADDRESS_MAP.containsKey(PUBLIC_NAMING_ADDRESS_PREFIX+clusterName)){
+                Service service = getNamingMaintainInstance().queryService(DEFAULT_APPLICATION,clusterName);
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:149]: meta data for every cluster id?

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 8]
```diff
@@ -129,6 +140,29 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
         if (clusterName == null) {
             return null;
         }
+        if (patternOfNacosRegistryForSLB == null) {
+            patternOfNacosRegistryForSLB = StringUtils.isBlank(FILE_CONFIG.getConfig(getNacosUrlPatternOfSLB()))
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:144]: duplicate get `FILE_CONFIG.getConfig(getNacosUrlPatternOfSLB()`

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 9]
```diff
@@ -39,6 +32,16 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.net.InetSocketAddress;
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:35]: reset java.* import header

**Review by slievrly — APPROVED** [review]
LGTM

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 10]
```diff
@@ -129,6 +140,30 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
         if (clusterName == null) {
             return null;
         }
+        if (patternOfNacosRegistryForSLB == null) {
```
Comment (a364176773) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:143]: 是否应该在初始化直接对patternOfNacosRegistryForSLB和useSLBWay处理？

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 11]
```diff
@@ -81,6 +81,13 @@ public class NacosRegistryServiceImpl implements RegistryService<EventListener>
     private static Boolean useSLBWay;
 
     private NacosRegistryServiceImpl() {
+        if (patternOfNacosRegistryForSLB == null) {
```
Comment (slievrly) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:84]: why judge if null,  it's a singleton.

File: seata-spring-autoconfigure/seata-spring-autoconfigure-core/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java (hunk 3 above)
Comment (2129zxl) [reply seata-spring-autoconfigure/seata-spring-autoconfigure-core/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java]: > 

this config can be auto added as it's part of nacos @ConfigurationProperties

**Comment by 2129zxl** [comment]
> Configuration item completion: https://github.com/seata/seata/blob/develop/seata-spring-autoconfigure/seata-spring-autoconfigure-client/src/main/resources/META-INF/additional-spring-configuration-metadata.json

this config can be auto added as it's part of nacos @ ConfigurationProperties

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 12]
```diff
@@ -61,15 +64,28 @@ public class NacosRegistryServiceImpl implements RegistryService<EventListener>
...
     private static volatile NacosRegistryServiceImpl instance;
+    private static volatile NamingMaintainService namingMaintain;
     private static final Object LOCK_OBJ = new Object();
+    private static final Pattern DEFAULT_SLB_REGISTRY_PATTERN = Pattern.compile("(?!.*internal)(?=.*seata).*mse.aliyuncs.com");
+    private static Pattern patternOfNacosRegistryForSLB;
+    private static Boolean useSLBWay;
 
     private NacosRegistryServiceImpl() {
+        String configForNacosSLB = FILE_CONFIG.getConfig(getNacosUrlPatternOfSLB());
+        patternOfNacosRegistryForSLB = StringUtils.isBlank(configForNacosSLB)
```
Comment (a364176773) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:85]: 为什么patternOfNacosRegistryForSLB不是个局部变量？

File: discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java [hunk 13]
```diff
@@ -61,15 +64,28 @@ public class NacosRegistryServiceImpl implements RegistryService<EventListener>
...
     private static final Configuration FILE_CONFIG = ConfigurationFactory.CURRENT_FILE_INSTANCE;
     private static volatile NamingService naming;
     private static final ConcurrentMap<String, List<EventListener>> LISTENER_SERVICE_MAP = new ConcurrentHashMap<>();
     private static final ConcurrentMap<String, List<InetSocketAddress>> CLUSTER_ADDRESS_MAP = new ConcurrentHashMap<>();
     private static volatile NacosRegistryServiceImpl instance;
+    private static volatile NamingMaintainService namingMaintain;
     private static final Object LOCK_OBJ = new Object();
+    private static final Pattern DEFAULT_SLB_REGISTRY_PATTERN = Pattern.compile("(?!.*internal)(?=.*seata).*mse.aliyuncs.com");
+    private static Pattern patternOfNacosRegistryForSLB;
+    private static Boolean useSLBWay;
```
Comment (a364176773) [inline discovery/seata-discovery-nacos/src/main/java/io/seata/discovery/registry/nacos/NacosRegistryServiceImpl.java:81]: private static volatile

**Review by a364176773 — APPROVED** [review]
LGTM

**Pull request merged**